numpy/ma/core.py

```diff
@@ -3709,6 +3724,11 @@ def filled(self, fill_value=None):
             fill_value = self.fill_value
         else:
             fill_value = _check_fill_value(fill_value, self.dtype)
+            if not (fill_value.ndim == 0 or self.shape == fill_value.shape):
+                warnings.warn(
+                    "Non-scalar arrays for fill values are deprecated. Use "
+                    "arrays with scalar values instead",
+                    DeprecationWarning, stacklevel=2)
```

Comment:
Test needed. It should go into `numpy/ma/tests/test_deprecations.py` as a new class `TestFill` and simply `assert_warns(DeprecationWarning, ...)`

Reply:
I added a new commit moving the deprecation (it was on the wrong spot). So now setting the `masked_array.fill_value = something` should give a deprecation warning, but the filled function is unchanged. So that is what we need to test.

There is still a problem that it has to be added at least in line 2904 in the `__new__` method. Can you make that line spell `self.fill_value = fill_value` and write a test that it works.

The indexing code seems like it might be slightly wrong, but I think it is unrelated (the warning says it makes it 0D which is what we want). (so please ignore this last part)